src/main.py

```diff
@@ -38,69 +39,147 @@ def get_current_metric_definitions():
-        f'{metric_name}-{six_days_before.strftime("%m-%d-%Y")}-to-{latest.strftime("%m-%d-%Y")}',
-        'percentageChange'
-    ]
 
-    ret_df = pd.DataFrame(ret, columns = column_names)
-    return ret_df
+def population_for_area(populations_df, area_name, area_code):
+    try:
+        area_population = populations_df.at[area_code, "All ages"]
+        return area_population
+    except KeyError:
+        print(f"Error getting population for area '{area_name}', area code {area_code} not found.", file=sys.stderr)
+
+
+def metric_value_per_100000(populations_df, metric_value, area_name, area_code):
+    area_population = population_for_area(populations_df, area_name, area_code)
+    return (metric_value / area_population) * 100000
+
+
+def get_percentage_change(area_name, metric_name, metric_df, aggregation_function):
+    area_data = metric_df[metric_df.areaName.eq(area_name)]
+
+    latest = metric_df['date'].max()
+
+    # We want 7 days inclusive of the latest
+    six_days_before = latest - pd.Timedelta(days=6)
+    seven_days_before = latest - pd.Timedelta(days=7)
+    # As above
+    thirteen_days_before = latest - pd.Timedelta(days=13)
+
+    area_data_last_week = area_data[area_data.date.ge(six_days_before)]
+    area_data_week_before_last = area_data[area_data.date.ge(thirteen_days_before) & area_data.date.le(seven_days_before)]
+
+    aggregation_output_last_week = getattr(area_data_last_week[metric_name], aggregation_function)()
+    aggregation_output_week_before = getattr(area_data_week_before_last[metric_name], aggregation_function)()
+
+    percentage_change = ((aggregation_output_last_week - aggregation_output_week_before) / aggregation_output_week_before) * 100.0
+    return {
+        "aggregation_output_week_before": {
+            "column_name": f'{metric_name}-{thirteen_days_before.strftime("%m-%d-%Y")}-to-{seven_days_before.strftime("%m-%d-%Y")}',
+            "value": aggregation_output_week_before
+        },
+        "aggregation_output_last_week": {
+            "column_name": f'{metric_name}-{six_days_before.strftime("%m-%d-%Y")}-to-{latest.strftime("%m-%d-%Y")}',
+            "value": aggregation_output_last_week
+        },
+        "percentage_change": {
+            "column_name": 'percentageChange',
+            "value": percentage_change
+        }
+    }
+
+def get_combined_stats(url, metric_name, aggregation_function):
+    metric_df = pd.read_csv(url)
+    metric_df['date'] = pd.to_datetime(metric_df['date'])
+
+    area_names = metric_df.areaName.unique()
+    ret = []
 
-def get_areas_above_threshold(area_type, metric_name, threshold, aggregationFunction):
+    if metric_name == "newCasesBySpecimenDate":
+        populations_df = get_populations()
+        for area_name in area_names:
+            percentage_change_stats = get_percentage_change(area_name, metric_name, metric_df, aggregation_function)
```

Comment:
I'd find this slightly more readable if calculating the stats for new cases and other metrics were broken into separate functions.

Reply:
superseded by comment below